connectors/memory/memory.go

```diff
@@ -278,28 +294,40 @@ func (c *Connector) Read(_ context.Context, ei *dosa.EntityInfo, values map[stri
 	return partitionRef[inx], nil
 }
 
+func overwriteValuesFunc(into map[string]dosa.FieldValue, from map[string]dosa.FieldValue) error {
+	for k, v := range from {
+		into[k] = v
+	}
+	return nil
+}
+
 // Upsert works a lot like CreateIfNotExists but merges the data when it finds an existing row
 func (c *Connector) Upsert(_ context.Context, ei *dosa.EntityInfo, values map[string]dosa.FieldValue) error {
-	return c.mergedInsert(ei.Def.Name, ei.Def.Key, values, func(into map[string]dosa.FieldValue, from map[string]dosa.FieldValue) error {
-		for k, v := range from {
-			into[k] = v
-		}
-		return nil
-	})
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if err := c.mergedInsert(ei.Def.Name, ei.Def.Key, values, overwriteValuesFunc); err != nil {
+		return err
+	}
+	for iName, iDef := range ei.Def.Indexes {
+		_ = c.mergedInsert(iName, ei.Def.UniqueKey(iDef.Key), values, overwriteValuesFunc)
```

Comment:
this also relies on the weird implementation details of `mergedInsert` but you forgot the doc here, making it unclear why we ignore the error. I think this is further evidence that we shouldn't call merged insert if we know it's going to fail. We should "document the behavior with code"